[scripts] add new function vcpkg_fixup_pkgconfig

Neumann-A opened the pull request:
similar to vcpkg_fixup_cmake_targets. 
Mainly prefix corrections and transforming absolute paths into relativ ones (relative to prefix). 

tries to solve #3108

Neumann-A committed:
add new function vcpkg_fixup_pkgconfig

vicroms commented on scripts/cmake/vcpkg_fixup_pkgconfig.cmake:
```diff
@@ -0,0 +1,49 @@
+    endif()
+    
+    message(STATUS "Fixing pkgconfig - release")
+    message(STATUS "Files: ${_vfpkg_RELEASE_FILES}")
```
We should minimize log output

Neumann-A replied:
there are cases where I wished there would be a status message in `vcpkg_fixup_cmake_targets` (vtk build; lot of cmakes files; vcpkg_fixup_cmake_targets takes a long time; build log file does not update anymore)
As such I made the second message a `debug_message` and kept the first one

vicroms replied:
Seems like vcpkg needs a `--verbose` flag

Neumann-A committed:
make message with changed files a debug_message

vicroms commented on scripts/cmake/vcpkg_fixup_pkgconfig.cmake:
```diff
@@ -0,0 +1,49 @@
+        file(READ "${_file}" _contents)
+        string(REPLACE "${CURRENT_PACKAGES_DIR}" "\${prefix}" _contents "${_contents}")
+        string(REPLACE "${CURRENT_INSTALLED_DIR}" "\${prefix}" _contents "${_contents}")
+        string(REGEX REPLACE "^prefix=\\\${prefix}" "prefix=${CURRENT_INSTALLED_DIR}" _contents "${_contents}")
```
Is there any way to make `pkgconfig` files relocatable? The hard-coded absolute paths don't work with scenarios like binary caching or export.

Some ideas:

- To consume `pkgconfig` files inside VCPKG ports, instead of fixing `.pc` files upon install, fix them before consumption.

    Example: In port `a` that needs `.pc` files from port `b`, `c` and `d`

    ```
    vcpkg_fixup_pkgconfig(b c d)
    ```

- To consume `pkgconfig` files outside of vcpkg (e.g. integrate with Meson), use a command that accepts an output location and triplet, and then extracts all `.pc` files  in `installed/<triplet>` folder to the output location while fixing the prefix paths.

    ```
    mkdir pkgconfig-files
    vcpkg x-get-pkgconfig-files x64-windows ./pkgconfig-files
    ```

Neumann-A replied:
> Is there any way to make pkgconfig files relocatable?

I reread the documentation at https://linux.die.net/man/1/pkg-config and it states for Windows:
>If a .pc file is found in a directory that matches the usual conventions (i.e., ends with \lib\pkgconfig), the prefix for that package is assumed to be the grandparent of the directory where the file was found, and the prefix variable is overridden for that file accordingly. 

as such any correction to the hardcoded prefix path is not needed. All other variables should be made relativ to prefix which is the main goal of this PR.

> To consume pkgconfig files inside VCPKG ports, instead of fixing .pc files upon install, fix them before consumption.

That would require touching N files for each single build. Seems unfeasiable and slow. Maybe vcpkg should/can itself detect if it has been reloacted and than perform the required actions. 

>To consume pkgconfig files outside of vcpkg (e.g. integrate with Meson), use a command that accepts an output location and triplet, and then extracts all .pc files in installed/<triplet> folder to the output location while fixing the prefix paths.

Currently I am more concerned about inside use than outside use. But there is always the option to call pkgconfig with `--define-variable=VARIABLENAME=VARIABLEVALUE` which can be used to overwrite prefix. The important part here is that all others PATHs in the *.pc files are relativ to prefix.

vicroms replied:
I see, making all paths relative to `prefix` is indeed desirable.

Then could we get away with not defining `prefix` in the file? On Windows I would expect `prefix` to be defined and on other platforms it would require the user to use `--define-variable`. Can you make that change?

Neumann-A replied:
> Then could we get away with not defining prefix in the file

Nope:
`configure: error: /../include/X11 doesn't exist or isn't a directory`
but with 
`set(ENV{PKG_CONFIG} "${PKGCONFIG} --define-variable=prefix=${_VCPKG_INSTALLED}")`
it seems to works in WSL using autoconf/configure (PKGCONFIG path setup with find_program)

but let me comment it instead of removing it completely since this way people are able to simply 'regex replace' it if they want to.

Neumann-A committed:
add two more cases to the debug *.pc files

Neumann-A committed:
comment out prefix.
use --define-variable=prefix=INSTALL_PATH

vicroms commented:
/azp run

Neumann-A commented:
@vicroms 
`set(ENV{PKG_CONFIG} "${PKGCONFIG} --define-variable=prefix=${_VCPKG_INSTALLED}")`
seems to be always required since 
`./pkg-config.exe --exists --print-errors "xproto"`
will always print 
`Variable 'prefix' not defined in 'd:/ma ke/installed/x64-windows/lib/pkgconfig\xproto.pc'` if there is no prefix variable in it. Otherwise the reseating of the prefix is done automatically on windows even when `prefix=nonsens` with pkgconfig v0.29.2
`pkg-config.exe --variable=prefix xproto -> d:/ma\ ke/installed/x64-windows`
while on linux the result is (v0.28.0)
`pkg-config.exe --variable=prefix xproto -> nonsens`

Neumann-A commented:
So my question now would be:
Should we let the prefix be defined on windows to leverage on the automatic prefix correction?
(while other OSes are required to use `--define-variable=prefix=${_VCPKG_INSTALLED}`)

vicroms commented:
**Questions:** On Windows, if there's no `prefix` defined in the file, does `pkgconfig` not define it automatically?  _We want to leverage the automatic correction, so if there's no other option we should define it._

I think that internally we should pass `--define-variable=prefix=${_VCPKG_INSTALLED}` in our helper functions regardless of platform.  For external consumers we could do something similar to the usage message we have for CMake config files and tell them to define the prefix.

ras0219-msft commented:
>`set(ENV{PKG_CONFIG} "${PKGCONFIG} --define-variable=prefix=${_VCPKG_INSTALLED}")`

+1 for this.

Just to reiterate the underlying sentiment and goal: we don't want packages to magically work on the original build machine, but randomly fail if you try to use them on another machine. Instead, we want them to fail consistently in both environments or work in both. If that requires users to throw `--define-variable=prefix=...`, so be it; we will find ways of making that more ergonomic.

ras0219-msft commented on scripts/cmake/vcpkg_fixup_pkgconfig.cmake:
```diff
@@ -0,0 +1,54 @@
+# .. command:: vcpkg_fixup_pkgconfig
+#
+# Tries to fix the paths found in *.pc files
+
```
As a note: we use a different comment style, which is extracted into `.md` files via `docs/regenerate.ps1`. See some other helpers like `vcpkg_configure_cmake()` for examples.

Neumann-A replied:
changed but still not sure if its 100% correct.

Neumann-A committed:
changed the comment header.

Neumann-A committed:
add missing word

Neumann-A commented:
I'll add in a check which will check if the following:
```
Libs: -L${libdir} -lfreetype
Libs.private: 
```
really exist. Translates -l<name> into lib<name>.(a|lib) or <name>.(a|lib) must exist. Most commonly, the debug name is false for CMake builds due to added debug suffixes while the library name is hardcoded in a *.pc.in

The same for 
```
Requires:
Requires.private: zlib, bzip2, libpng
```
which means that a <name>.pc file needs to exist

Neumann-A committed:
Merge branch 'master' into add_vcpkg_fixup_pkgconfig

Neumann-A committed:
finish vcpkg_fixup_pkgconfig.cmake

Voskrese commented:
@dan-shaw @ras0219-msft 
merge!
in the course of work, any shortcomings will be identified, need to merge
the script does not yet carry anything without accessing it

Voskrese commented on scripts/cmake/vcpkg_fixup_pkgconfig.cmake:
```diff
@@ -0,0 +1,261 @@
+    endforeach()
+    message(STATUS "Fixing pkgconfig --- finished")
+    
+    set(VCPKG_FIXUP_PKGCONFIG_CALLED TRUE CACHE INTERNAL FORCE) 
```
```suggestion
    set(VCPKG_FIXUP_PKGCONFIG_CALLED TRUE CACHE INTERNAL "Description goes here." FORCE) 
```

Voskrese replied:
-- Fixing pkgconfig --- finished
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:245 (set):
  set given invalid arguments for CACHE mode.

MVoz reviewed (commented):
-- Fixing pkgconfig --- finished
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:245 (set):
  set given invalid arguments for CACHE mode.

Neumann-A committed:
Update vcpkg_fixup_pkgconfig.cmake

Note: since CMake is run in script mode the description of VCPKG_FIXUP_PKGCONFIG_CALLED cannot be viewed

Voskrese commented:
> when there are dependencies, the script does not work, and outputs an error without pointing to the package.
>  I performed validation of all files

Neumann-A commented:
Could you add the `*.pc` file which failed?
I am not going to comment `vcpkg_fixup_pkgconfig_check_libraries` out since its intention is too make sure that all required packages and libraries are available.

Voskrese commented:
> **I am not suggesting that you comment on this line \ function**
> I removed it under the comment (at itself), so that the build was successful until I find out the reason for the error

Neumann-A commented:
> until I find out the reason for the error
> an error without pointing to the package.

So the reason for the error was that one package was missing but the error did not point to the missing package? Or am I misunderstanding it again?

Voskrese commented:
all packages were found and the build was made successfully, the script after the build can not work out some system files
which ones, I have no idea, I got the error without explanation

> пакеты все найдены и сборка была произведена успешно, скрипт после сборки не может отработать некие системные файлы
> какие именно, я без понятия, вывела ошибку без пояснения

Neumann-A committed:
transfer changes from x windows pr

JackBoosY commented:
/azp run

Neumann-A committed:
fix typo in regex

Neumann-A commented:
TODO:
https://github.com/microsoft/vcpkg/pull/10731#discussion_r405455495
Make prefix use `${pcfiledir}/../..`

Neumann-A committed:
make the regex comment aware
make pc files relocatable by using ${pcfiledir} in prefix
tested with x window pr

JackBoosY commented:
LGTM.
@ras0219-msft  Could you review this PR?

Thanks.

Voskrese commented:
@Neumann-A 
still would such a feature to implement, naturally in another request

example

# Templates release\debug.pc.cmake \\ RELEASE and DEBUG

prefix=@CMAKE_INSTALL_PREFIX@ ### @CURRENT_INSTALLED_DIR@ ???
exec_prefix=${prefix}
libdir=${exec_prefix}/@LIB_INSTALL_DIR@
includedir=${prefix}/include ### ${exec_prefix}/@INCLUDE_INSTALL_DIR@ ???

Name: @PORT@
URL: @PORT_HOMEPAGE@
Description: @PORT_DESCRIPTION@ ### error /n
Version: @PORT_VERSION@ ### ???
Requires: @PORT_REQUIRES_RELEASE@
Version: @EMBREE_VERSION@

Libs: -L${libdir} @PORT_LIBS_RELEASE@ ### -lexample
Cflags: -I${includedir}


# Generate and Install Pkgconfig

configure_file(common/cmake/release.pc.cmake ${PROJECT_BINARY_DIR}//@PORT@.pc @ONLY) ### ${PROJECT_BINARY_DIR} ???
install(FILES ${PROJECT_BINARY_DIR}/@PORT@.pc DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig") ### ${CMAKE_INSTALL_LIBDIR} ???

Microsoft-lgtm commented:
Pending





<!-- Sent from Reviewable.io -->

Voskrese commented:
@ras0219-msft merge this PR

Thanks.

strega-nil commented:
LGTM! Thanks @Neumann-A :)
